### PR TITLE
Add a way to exclude tests by specifying an exclusion file

### PR DESCRIPTION
**Pull request by MarekVCodasip**
This patch adds a way to specify a yaml file which specifies either for each target individually or for all targets to exclude tests. Example file format is included in excluded.yaml.example.

### CODE REVIEW

**Comment by MarekVCodasip** [comment]
Hey @timsifive, can you please take a look at this.

File: debug/excluded.yaml.example [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+# Tests excluded for the "spike32" target only
+spike32:
+  - MemTest32
+  - PrivRw
+  - Sv32Test
+
+# Tests excluded for the "spike64" target only
+spike64:
+  - UserInterrupt
```
Comment (JanMatCodasip) [inline debug/excluded.yaml.example:17]: @MarekVCodasip Detail - it looks like there's a missing newline at the end of this text file.

(Explanation: https://stackoverflow.com/a/729795)

**Review by JanMatCodasip — COMMENTED** [review]
@MarekVCodasip One more improvement suggestion: You can also add the following to the Makefile to simplify the use of the exclusion lists:

```diff
diff --git a/debug/Makefile b/debug/Makefile
index 06f7d9d..b412e58 100644
--- a/debug/Makefile
+++ b/debug/Makefile
@@ -23,7 +23,8 @@ run.%:
                --isolate \
                --print-failures \
                --sim_cmd $(RISCV)/bin/$(RISCV_SIM) \
-               --server_cmd $(RISCV)/bin/openocd
+               --server_cmd $(RISCV)/bin/openocd \
+               $(if $(EXCLUDE_TESTS),--exclude-tests $(EXCLUDE_TESTS))
 
 # Target to check all the multicore options.
 multi-tests: spike32-2 spike32-2-hwthread

```

The use would then be:

```bash
EXCLUDE_TESTS=my_exclusion_list.yml make all
```

File: debug/excluded.yaml.example (hunk 1 above)
Comment (MarekVCodasip) [reply debug/excluded.yaml.example]: Added

File: debug/testlib.py [hunk 2]
```diff
@@ -947,6 +948,31 @@ def __enter__(self):
...
+        return result
+
+    target_excludes = {}
+    with open(excluded_tests_file) as file:
+        raw_data = yaml.safe_load(file)
+        for test_list in raw_data.values():
+            if not isinstance(test_list, list):
+                raise ValueError(excluded_tests_file, test_list)
+            if not all(isinstance(s, str) for s in test_list):
+                raise ValueError(excluded_tests_file)
```
Comment (timsifive) [inline debug/testlib.py:964]: Can we provide the user a meaningful error message here?

**Review by timsifive — CHANGES_REQUESTED** [review]
Looks like a good feature, that could be used in the future to maintain known lists of failing tests as well.

**Review by JanMatCodasip — APPROVED** [review]

File: debug/testlib.py [hunk 3]
```diff
@@ -947,6 +948,33 @@ def __enter__(self):
...
+        return result
+
+    target_excludes = {}
+    with open(excluded_tests_file) as file:
+        raw_data = yaml.safe_load(file)
+        invalid_format_message = "Invalid format for the exclusion file. Please take a look at the " \
+                                 "\"excluded.yaml.example\" file as an example of the test exclusion file format."
+        for test_list in raw_data.values():
+            if not isinstance(test_list, list):
+                raise ValueError(invalid_format_message, excluded_tests_file, test_list)
```
Comment (timsifive) [inline debug/testlib.py:964]: Specifically here, can you tell the user something like "list of tests wasn't a list".

Even better would be to do `for (target, test_list) in raw_data.items()` above so the error can include the name of the target the list belongs to.

File: debug/testlib.py [hunk 4]
```diff
@@ -947,6 +948,33 @@ def __enter__(self):
...
+            if not all(isinstance(s, str) for s in test_list):
+                raise ValueError(invalid_format_message, excluded_tests_file)
```
Comment (timsifive) [inline debug/testlib.py:966]: And here, "not every element in the test list is a string"

**Review by timsifive — CHANGES_REQUESTED** [review]

File: debug/testlib.py (hunk 3 above)
Comment (MarekVCodasip) [reply debug/testlib.py]: Done

File: debug/testlib.py (hunk 4 above)
Comment (MarekVCodasip) [reply debug/testlib.py]: Added with the target name

File: debug/testlib.py [hunk 5]
```diff
@@ -947,6 +948,32 @@ def __enter__(self):
...
+    result = []
+    if excluded_tests_file is None or len(excluded_tests_file) == 0:
+        return result
+
+    target_excludes = {}
+    with open(excluded_tests_file) as file:
+        raw_data = yaml.safe_load(file)
+        for (target, test_list) in raw_data.items():
+            if not isinstance(test_list, list):
+                raise ValueError(f"Target \"{target}\" does not contain a test list", excluded_tests_file, test_list)
```
Comment (timsifive) [inline debug/testlib.py:962]:
```suggestion
                raise ValueError(f"Target {target!r} does not contain a test list", excluded_tests_file, test_list)
```

File: debug/testlib.py [hunk 6]
```diff
@@ -947,6 +948,32 @@ def __enter__(self):
...
+            if not all(isinstance(s, str) for s in test_list):
+                raise ValueError(f"Not every element in the target test list \"{target}\" is a string",
```
Comment (timsifive) [inline debug/testlib.py:964]:
```suggestion
                raise ValueError(f"Not every element in the target test list {target!r} is a string",
```

File: debug/testlib.py (hunk 5 above)
Comment (timsifive) [reply debug/testlib.py:962]: This way python will add the quotes if it is a string, and also escape newlines or whatever else might end up in the string if the data is malformed.

**Review by timsifive — CHANGES_REQUESTED** [review]
Last nitpicky comments.

**Review by timsifive — APPROVED** [review]

**Pull request merged**